File: solver/internal/solver/solver.go
// Package solver implements the exact minimum-purchase optimization.
package solver

import (
	"context"
	"errors"

	"github.com/arceoavs/papsolver/solver/internal/domain"
)

// ErrNoExactSolution means no nonnegative combination reaches the target.
var ErrNoExactSolution = errors.New("no exact solution")

type Assignment struct {
	TierID     string
	Label      *string
	PriceCents int64
	Quantity   int
}

type Result struct {
	TargetCents   int64
	PurchaseCount int
	Assignments   []Assignment
}

type activeTier struct {
	id              string
	label           *string
	priceCents      int64
	normalizedPrice int
}

// Solve returns a deterministic exact solution with the fewest purchases.
// Request order breaks ties between solutions with the same purchase count.
func Solve(ctx context.Context, problem domain.Problem) (Result, error) {
	if err := ctx.Err(); err != nil {
		return Result{}, err
	}

	targetCents := problem.Target().MinorUnits()
	tiers := problem.Tiers()
	active := make([]activeTier, 0, len(tiers))
	var divisor int64
	for _, tier := range tiers {
		price := tier.Price().MinorUnits()
		if price > targetCents {
			continue
		}
		var label *string
		if value, ok := tier.Label(); ok {
			label = &value
		}
		divisor = greatestCommonDivisor(divisor, price)
		active = append(active, activeTier{id: tier.ID(), label: label, priceCents: price})
	}
	if len(active) == 0 || targetCents%divisor != 0 {
		return Result{}, ErrNoExactSolution
	}

	normalizedTarget := int(targetCents / divisor)
	for index := range active {
		active[index].normalizedPrice = int(active[index].priceCents / divisor)
	}

	unreachable := normalizedTarget + 1
	purchaseCounts := make([]int, normalizedTarget+1)
	selectedTier := make([]int, normalizedTarget+1)
	for subtotal := 1; subtotal <= normalizedTarget; subtotal++ {
		purchaseCounts[subtotal] = unreachable
		selectedTier[subtotal] = -1
	}

	for subtotal := 1; subtotal <= normalizedTarget; subtotal++ {
		if subtotal&255 == 0 {
			if err := ctx.Err(); err != nil {
				return Result{}, err
			}
		}

		bestCount := unreachable
		bestTier := -1
		for tierIndex, tier := range active {
			previous := subtotal - tier.normalizedPrice
			if previous < 0 || purchaseCounts[previous] == unreachable {
				continue
			}
			candidate := purchaseCounts[previous] + 1
			if candidate < bestCount {
				bestCount = candidate
				bestTier = tierIndex
			}
		}
		purchaseCounts[subtotal] = bestCount
		selectedTier[subtotal] = bestTier
	}

	if selectedTier[normalizedTarget] == -1 {
		return Result{}, ErrNoExactSolution
	}

	quantities := make([]int, len(active))
	for subtotal := normalizedTarget; subtotal > 0; {
		tierIndex := selectedTier[subtotal]
		quantities[tierIndex]++
		subtotal -= active[tierIndex].normalizedPrice
	}

	assignments := make([]Assignment, 0, len(active))
	for index, quantity := range quantities {
		if quantity == 0 {
			continue
		}
		assignments = append(assignments, Assignment{
			TierID:     active[index].id,
			Label:      active[index].label,
			PriceCents: active[index].priceCents,
			Quantity:   quantity,
		})
	}

	return Result{
		TargetCents:   targetCents,
		PurchaseCount: purchaseCounts[normalizedTarget],
		Assignments:   assignments,
	}, nil
}

func greatestCommonDivisor(left, right int64) int64 {
	for right != 0 {
		left, right = right, left%right
	}
	return left
}

File: solver/app/solver.py
from collections.abc import Sequence
from functools import reduce
from math import gcd

from .models import Tier


def solve_exact(tiers: Sequence[Tier], target: int) -> list[int] | None:
    """Return a deterministic minimum-purchase assignment, or ``None``.

    The tier order is the tie-breaker: when equally small assignments exist, the
    first tier in the request is preferred at the first differing choice.
    """
    active_tiers = [
        (tier_index, tier.price.int)
        for tier_index, tier in enumerate(tiers)
        if tier.price.int <= target
    ]
    if not active_tiers:
        return None

    prices = [price for _, price in active_tiers]
    common_divisor = reduce(gcd, prices)
    if target % common_divisor:
        return None

    normalized_target = target // common_divisor
    normalized_prices = [price // common_divisor for price in prices]
    unreachable = normalized_target + 1
    purchase_counts = [unreachable] * (normalized_target + 1)
    selected_tier = [-1] * (normalized_target + 1)
    purchase_counts[0] = 0

    for subtotal in range(1, normalized_target + 1):
        best_count = unreachable
        best_tier = -1
        for tier_index, price in enumerate(normalized_prices):
            previous = subtotal - price
            if previous < 0 or purchase_counts[previous] == unreachable:
                continue

            candidate = purchase_counts[previous] + 1
            if candidate < best_count:
                best_count = candidate
                best_tier = tier_index

        purchase_counts[subtotal] = best_count
        selected_tier[subtotal] = best_tier

    if selected_tier[normalized_target] == -1:
        return None

    assignment = [0] * len(tiers)
    subtotal = normalized_target
    while subtotal:
        active_tier_index = selected_tier[subtotal]
        original_tier_index = active_tiers[active_tier_index][0]
        assignment[original_tier_index] += 1
        subtotal -= normalized_prices[active_tier_index]

    return assignment
